### DISCUSSION TITLE
Two `RestMulti` enhancement ideas

### DISCUSSION THREAD

**Discussion by snazy** [resteasy/resteasy#4032] [Ideas]
I have two proposals to enhance `RestMulti`.

1. Produce multiple JSON objects, not just an array
2. Make requested demand configurable

I could contribute the necessary code changes.

## Produce multiple JSON objects

Currently, `PublisherResponseHandler.StreamingMultiSubscriber` produces a JSON array, where each emitted item is encoded as a JSON array element. For some use cases it is easier to consume a bunch of "bare" JSON objects - i.e. just write the individual JSON objects, possibly separated by a newline. As an option, of course.

Proposal to add:
```java
RestMulti.fromMultiData(multi).encodeAsArray(false)...
```

With `encodeAsArray(false)`, the produced JSON would look like this:
```json
{"some": "value"},
{"some": "value"},
{"some": "value"}
```

`encodeAsArray(true)` or omitting it would use the current behavior and produce something like this:
```json
[{"some": "value"},
{"some": "value"},
{"some": "value"}
}
```

## Configure request-demand

All implementations of `PublisherResponseHandler.AbstractMultiSubscriber` work with a hard-coded request-demand of `1`, which means that every emitted item is "produced"/"computed" serially / one-after-the-other. If the computation of individual items takes somewhat longer, possibly waiting for remote resources to reply, it makes sense to use a higher demand to produce multiple items concurrently.

For example, if each item takes maybe 250 ms (requesting data from a remote source) to be produced, and 100 items are produced, it currently takes 25 seconds. With a higher concurrency it would take a fraction of that time. I.e. if the use case is known to be not CPU but (async) I/O bound, it _might_ be legit/feasible to use a high demand.

Proposal to add:
```
RestMulti.fromMultiData(multi).withDemand( (long) 123 )...
```

Which would pass `123` as the demand for all call sites to `Subscription.request` in implementations of `PublisherResponseHandler.AbstractMultiSubscriber`.

**Comment by jamezp** [comment]
I'm guessing this should be for RESTEasy Reactive in Quarkus as RESTEasy doesn't have a `RestMulti` type.

**Comment by snazy** [comment]
Oh, that's actually true. I was probably misguided by the package name.
/cc @geoand WDYT?

**Comment by geoand** [comment]
@snazy sounds good!

Would you like to contribute these?

**Reply by snazy** [reply]
Sure! Let me open a PR against Quarkus in a bit.

**Reply by geoand** [reply]
💘

**Comment by snazy** [comment]
Closing as it's only related to Quarkus.
